Fix nan_opacity issues and above/below annotations

banesullivan opened the pull request:
Resolves #3445 (see comments there)

@dcbr, a review would be appreciated.

dcbr reviewed (approved):
LGTM if my comment in #3445 is correct, i.e.

> So I believe for this use case (add_mesh) the usage of opacity is correct. You want to override any alpha values of the provided (or theme) nan_color.

banesullivan commented:
Thanks for the feedback, @dcbr!

I also just added some tests to verify `nan_opacity` works.

@akaszynski, I modified `LoolupTable` to have a `nan_opacity` property, but this feels a bit hacky since `nan_opacity` is really a part of `nan_color`. I feel like my change might have some consequence... would you please review?

banesullivan commented on tests/plotting/test_plotting.py:
```diff
@@ -3031,13 +3031,28 @@ def test_lookup_table(verify_image_cache):
+    # NaN values should be hidden
+    pl = pyvista.Plotter()
+    pl.add_mesh(uniform, nan_opacity=0)
+    pl.enable_depth_peeling()
```
Depth peeling makes the opacity of the rest of the mesh correct... is this okay to include for CI tests?

dcbr commented on pyvista/plotting/lookup_table.py:
```diff
@@ -513,6 +514,31 @@ def nan_color(self, value):
+
+    @nan_opacity.setter
+    def nan_opacity(self, value):
+        nan_color = self.nan_color.float_rgba[:3]
+        self.nan_color = Color(nan_color, opacity=value).float_rgba
```
```suggestion
        self.nan_color = Color(self.nan_color, opacity=value)
```

Colors can be created from other `Color` instances, so no need to first extract the RGB components yourself. I also removed the conversion to `float_rgba` as I believe `self.nan_color` should be a `Color` instance?

akaszynski reviewed (dismissed):
This is good to go. Feel free to add to it and I'll reapprove.

banesullivan commented:
Made a few more changes to ensure that if you build a LookupTable with above/below/nan color annotations that the scalar bar would be set up correctly

banesullivan commented:
NaN must be explicitly requested for the scalar bar though as the NaN color is always set on the LookupTable and we can't really track when it's user set vs. default set without some messy code.

To get a nan-annotaion, users must:

```py
pl.add_mesh(
        ..., nan_opacity=0.5, nan_color='green', scalar_bar_args=dict(nan_annotation=True)
    )
```

dcbr commented on pyvista/plotting/colors.py:
```diff
@@ -872,6 +872,33 @@ def to_dict(self):
+        """
+        return self._opacity
+
+    @opacity.setter
+    def opacity(self, value: Union[int, float, str]):
+        self._opacity = self.convert_color_channel(value)
```
I decided against making Color instances mutable in #2175. The reason being that sometimes colors are only used as "intermediate objects" between the Pyvista user and the underlying VTK methods. See e.g.

https://github.com/pyvista/pyvista/blob/fecf347ca56221c3617a0121ec2cd603b9378439/pyvista/plotting/charts.py#L196-L216

If a user uses the color getter of a `Pen` instance and modifies the opacity, the `Color` instance will be changed, but the `SetColor` VTK method won't be called to actually update the color (as would be the case when using the color setter).
This would also break the hashing of `Color` instances (hash value is calculated from each color channel).

banesullivan replied:
ah makes sense, thanks for pointing this out. I'll modify this PR to make sure it is not mutating any `Color` objects

banesullivan commented on pyvista/plotting/scalar_bars.py:
```diff
@@ -164,7 +164,7 @@ def add_scalar_bar(
         use_opacity=True,
         outline=False,
-        nan_annotation=False,
+        nan_annotation=None,
```
Maybe this should be kept `False`

akaszynski reviewed (approved):
LGTM. Thanks @banesullivan!

Please merge and ignore missing codecov, it's a server-side failure to report, not our fault.

akaszynski commented:
Merging this immediately as this is a high priority and it's been sitting here for a while.
